
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>analysis Module &#8212; phyphy  documentation</title>
    <link rel="stylesheet" href="_static/haiku.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="extractor Module" href="extractor.html" />
    <link rel="prev" title="hyphy Module" href="hyphy.html" /> 
  </head><body>
      <div class="header" role="banner"><h1 class="heading"><a href="index.html">
          <span>phyphy  documentation</span></a></h1>
        <h2 class="heading"><span>analysis Module</span></h2>
      </div>
      <div class="topnav" role="navigation" aria-label="top navigation">
      
        <p>
        «&#160;&#160;<a href="hyphy.html"><code class="docutils literal notranslate"><span class="pre">hyphy</span></code> Module</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="extractor.html"><code class="docutils literal notranslate"><span class="pre">extractor</span></code> Module</a>&#160;&#160;»
        </p>

      </div>
      <div class="content">
        
        
  <div class="section" id="module-analysis">
<span id="analysis-module"></span><h1><code class="docutils literal notranslate"><span class="pre">analysis</span></code> Module<a class="headerlink" href="#module-analysis" title="Permalink to this headline">¶</a></h1>
<p>Execute a standard HyPhy analysis.</p>
<dl class="class">
<dt id="analysis.Analysis">
<em class="property">class </em><code class="descclassname">analysis.</code><code class="descname">Analysis</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#analysis.Analysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Parent class for all analysis methods, which include the following children:</p>
<ul class="simple">
<li>ABSREL</li>
<li>BUSTED</li>
<li>FEL</li>
<li>FUBAR</li>
<li>LEISR</li>
<li>MEME</li>
<li>RELAX</li>
<li>SLAC</li>
</ul>
<p><strong>Do not use this parent class. Instead, see child classes for analysis-specific arguments and examples.</strong></p>
<dl class="method">
<dt id="analysis.Analysis.run_analysis">
<code class="descname">run_analysis</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#analysis.Analysis.run_analysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Execute an Analysis and save output.</p>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Execute a default FEL analysis</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfel</span> <span class="o">=</span> <span class="n">FEL</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfel</span><span class="o">.</span><span class="n">run_analysis</span><span class="p">()</span>         
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="analysis.FEL">
<em class="property">class </em><code class="descclassname">analysis.</code><code class="descname">FEL</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#analysis.FEL" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#analysis.Analysis" title="analysis.Analysis"><code class="xref py py-class docutils literal notranslate"><span class="pre">analysis.Analysis</span></code></a></p>
<p>Initialize and execute a FEL analysis.</p>
<dl class="docutils">
<dt>Required arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>alignment</strong> and <strong>tree</strong> OR <strong>data</strong>, either a file for alignment and tree separately, OR a file with both (combo FASTA/newick or nexus). Note that if a NEXUS file is provided with the <cite>data</cite> argument, the additional argument <cite>nexus=True</cite> must be supplied.</li>
</ol>
</dd>
<dt>Optional keyword arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>hyphy</strong>, a <code class="code docutils literal notranslate"><span class="pre">HyPhy()</span></code> instance. Default: Assumes canonical HyPhy install.</li>
<li><strong>srv</strong>, Employ synonymous rate variation in inference (i.e. allow dS to vary across sites?). Values “Yes”/”No” or True/False accepted. Default: True.</li>
<li><strong>branches</strong>, Branches to consider in site-level selection inference. Values “All”, “Internal”, “Leaves”, “Unlabeled branches”, or a <strong>specific label</strong> in your tree are accepted</li>
<li><strong>output</strong>, Name (and path to) to final output JSON file. Default: Goes to same directory as provided data</li>
<li><strong>alpha</strong>, The p-value threshold for calling sites as positively or negatively selected. Note that this argument has 0 bearing on JSON output. Default: 0.1</li>
<li><strong>genetic_code</strong>, the genetic code to use in codon analysis, Default: Universal. Consult NIH for details.</li>
<li><strong>nexus</strong>, a Boolean <em>only required when</em> a nexus file is provided to the argument <cite>data</cite>. Default: False.</li>
</ol>
</dd>
</dl>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default FEL analysis, where data is contained in a single file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfel</span> <span class="o">=</span> <span class="n">FEL</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default FEL analysis, where data is contained in a single NEXUS file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfel</span> <span class="o">=</span> <span class="n">FEL</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.nex&quot;</span><span class="p">,</span> <span class="n">nexus</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default FEL analysis, where alignment and tree are in separate files </span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfel</span> <span class="o">=</span> <span class="n">FEL</span><span class="p">(</span><span class="n">alignment</span> <span class="o">=</span> <span class="s2">&quot;/path/to/alignment.fasta&quot;</span><span class="p">,</span> <span class="n">tree</span> <span class="o">=</span> <span class="s2">&quot;/path/to/tree.tre&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a FEL analysis, with a specified path to output JSON</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfel</span> <span class="o">=</span> <span class="n">FEL</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">output</span><span class="o">=</span><span class="s2">&quot;/path/to/json/output.json&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a FEL analysis, with a one-rate approach (i.e. synonymous rate variation turned off) </span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfel</span> <span class="o">=</span> <span class="n">FEL</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">srv</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define FEL analysis, specifying only to use internal branches to test for selection</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfel</span> <span class="o">=</span> <span class="n">FEL</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">branches</span><span class="o">=</span><span class="s2">&quot;Internal&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define FEL analysis with a custom Hyphy, which is also defined here:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myhyphy</span> <span class="o">=</span> <span class="n">HyPhy</span><span class="p">(</span><span class="n">suppress_log</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">quiet</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="c1">## HyPhy will use default canonical install but run in full quiet mode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfel</span> <span class="o">=</span> <span class="n">FEL</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">hyphy</span><span class="o">=</span><span class="n">myhyphy</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Execute a defined FEL instance</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfel</span><span class="o">.</span><span class="n">run_analysis</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="analysis.FUBAR">
<em class="property">class </em><code class="descclassname">analysis.</code><code class="descname">FUBAR</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#analysis.FUBAR" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#analysis.Analysis" title="analysis.Analysis"><code class="xref py py-class docutils literal notranslate"><span class="pre">analysis.Analysis</span></code></a></p>
<p>Initialize and execute a FUBAR analysis.</p>
<dl class="docutils">
<dt>Required arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>alignment</strong> and <strong>tree</strong> OR <strong>data</strong>, either a file for alignment and tree separately, OR a file with both (combo FASTA/newick or nexus). Note that if a NEXUS file is provided with the <cite>data</cite> argument, the additional argument <cite>nexus=True</cite> must be supplied.</li>
</ol>
</dd>
<dt>Optional keyword arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>hyphy</strong>, a <code class="code docutils literal notranslate"><span class="pre">HyPhy()</span></code> instance. Default: Assumes canonical HyPhy install.</li>
<li><strong>output</strong>, Name (and path to) to final output JSON file. Default: Goes to same directory as provided data.</li>
<li><strong>genetic_code</strong>, the genetic code to use in codon analysis, Default: Universal. Consult NIH for details.</li>
<li><strong>grid_size</strong>, Number of grid points per rate grid dimension (Default: 20, allowed [5,50])</li>
<li><strong>nchains</strong>, Number of MCMC chains to run (Default: 5, allowed [2,20])</li>
<li><strong>chain_length</strong>, The length of each chain (Default: 2e6, allowed [5e5,5e7])</li>
<li><strong>burnin</strong>, Number of samples to consider as burn-in (Default 1e6, allowed [ceil(chain_length/20),ceil(95*chain_length/100)])</li>
<li><strong>samples_per_chain</strong>, Number of samples to draw per chain (Default 100, allowed [50,chain_length-burnin])</li>
<li><strong>alpha</strong>, The concentration parameter of the Dirichlet prior (Default 0.5, allowed[0.001,1])</li>
<li><strong>cache</strong>, Name (and path to) output FUBAR cache. Default: goes to same directory as provided data. Provide the argument <strong>False</strong> to not save the cache (this argument simply sends it to /dev/null)</li>
<li><strong>nexus</strong>, a Boolean <em>only required when</em> a nexus file is provided to the argument <cite>data</cite>. Default: False.</li>
<li><strong>method</strong>, The algorithm used for FUBAR posterior estimation, in FUBAR version &gt;=2.1. Default: “VB”. Options include “MH” and “CG”.</li>
</ol>
</dd>
</dl>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default FUBAR analysis, where data is contained in a single file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfubar</span> <span class="o">=</span> <span class="n">FUBAR</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default FUBAR analysis, where data is contained in a single NEXUS file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfUBAR</span> <span class="o">=</span> <span class="n">FUBAR</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.nex&quot;</span><span class="p">,</span> <span class="n">nexus</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default FUBAR analysis, where alignment and tree are in separate files </span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfubar</span> <span class="o">=</span> <span class="n">FUBAR</span><span class="p">(</span><span class="n">alignment</span> <span class="o">=</span> <span class="s2">&quot;/path/to/alignment.fasta&quot;</span><span class="p">,</span> <span class="n">tree</span> <span class="o">=</span> <span class="s2">&quot;/path/to/tree.tre&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a FUBAR analysis using a 10x10 grid and alpha parameter of 0.75</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfubar</span> <span class="o">=</span> <span class="n">FUBAR</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">grid_size</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.75</span> <span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define FUBAR analysis with a custom Hyphy, which is also defined here:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myhyphy</span> <span class="o">=</span> <span class="n">HyPhy</span><span class="p">(</span><span class="n">suppress_log</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">quiet</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="c1">## HyPhy will use default canonical install but run in full quiet mode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfubar</span> <span class="o">=</span> <span class="n">FUBAR</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">hyphy</span><span class="o">=</span><span class="n">myhyphy</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Execute a defined FUBAR instance</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myfubar</span><span class="o">.</span><span class="n">run_analysis</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="analysis.MEME">
<em class="property">class </em><code class="descclassname">analysis.</code><code class="descname">MEME</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#analysis.MEME" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#analysis.Analysis" title="analysis.Analysis"><code class="xref py py-class docutils literal notranslate"><span class="pre">analysis.Analysis</span></code></a></p>
<dl class="docutils">
<dt>Required arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>alignment</strong> and <strong>tree</strong> OR <strong>data</strong>, either a file for alignment and tree separately, OR a file with both (combo FASTA/newick or nexus). Note that if a NEXUS file is provided with the <cite>data</cite> argument, the additional argument <cite>nexus=True</cite> must be supplied.</li>
</ol>
</dd>
<dt>Optional keyword arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>hyphy</strong>, a <code class="code docutils literal notranslate"><span class="pre">HyPhy()</span></code> instance. Default: Assumes canonical HyPhy install.</li>
<li><strong>branches</strong>, Branches to consider in site-level selection inference. Values “All”, “Internal”, “Leaves”, “Unlabeled branches”, or a <strong>specific label</strong> are accepted</li>
<li><strong>output</strong>, Name (and path to) to final output JSON file. Default: Goes to same directory as provided data</li>
<li><strong>alpha</strong>, The p-value threshold for calling sites as positively or negatively selected. Note that this argument has 0 bearing on JSON output. Default: 0.1</li>
<li><strong>genetic_code</strong>, the genetic code to use in codon analysis, Default: Universal. Consult NIH for details.</li>
<li><strong>nexus</strong>, a Boolean <em>only required when</em> a nexus file is provided to the argument <cite>data</cite>. Default: False.</li>
</ol>
</dd>
</dl>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default MEME analysis, where data is contained in a single file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mymeme</span> <span class="o">=</span> <span class="n">MEME</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default MEME analysis, where data is contained in a single NEXUS file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mymeme</span> <span class="o">=</span> <span class="n">MEME</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.nex&quot;</span><span class="p">,</span> <span class="n">nexus</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default MEME analysis, where alignment and tree are in separate files </span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mymeme</span> <span class="o">=</span> <span class="n">MEME</span><span class="p">(</span><span class="n">alignment</span> <span class="o">=</span> <span class="s2">&quot;/path/to/alignment.fasta&quot;</span><span class="p">,</span> <span class="n">tree</span> <span class="o">=</span> <span class="s2">&quot;/path/to/tree.tre&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a MEME analysis, specifying that selection be tested only on leaves</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mymeme</span> <span class="o">=</span> <span class="n">MEME</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">branches</span> <span class="o">=</span> <span class="s2">&quot;Leaves&quot;</span> <span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a MEME analysis, specifying a custom JSON output file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mymeme</span> <span class="o">=</span> <span class="n">MEME</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;meme.json&quot;</span> <span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define MEME analysis with a custom Hyphy, which is also defined here:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myhyphy</span> <span class="o">=</span> <span class="n">HyPhy</span><span class="p">(</span><span class="n">suppress_log</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">quiet</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="c1">## HyPhy will use default canonical install but run in full quiet mode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mymeme</span> <span class="o">=</span> <span class="n">MEME</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">hyphy</span><span class="o">=</span><span class="n">myhyphy</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Execute a defined MEME instance</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mymeme</span><span class="o">.</span><span class="n">run_analysis</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="analysis.SLAC">
<em class="property">class </em><code class="descclassname">analysis.</code><code class="descname">SLAC</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#analysis.SLAC" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#analysis.Analysis" title="analysis.Analysis"><code class="xref py py-class docutils literal notranslate"><span class="pre">analysis.Analysis</span></code></a></p>
<dl class="docutils">
<dt>Required arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>alignment</strong> and <strong>tree</strong> OR <strong>data</strong>, either a file for alignment and tree separately, OR a file with both (combo FASTA/newick or nexus). Note that if a NEXUS file is provided with the <cite>data</cite> argument, the additional argument <cite>nexus=True</cite> must be supplied.</li>
</ol>
</dd>
<dt>Optional keyword arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>hyphy</strong>, a <code class="code docutils literal notranslate"><span class="pre">HyPhy()</span></code> instance. Default: Assumes canonical HyPhy install.</li>
<li><strong>branches</strong>, Branches to consider in site-level selection inference. Values “All”, “Internal”, “Leaves”, “Unlabeled branches”, or a <strong>specific label</strong> are accepted</li>
<li><strong>output</strong>, Name (and path to) to final output JSON file. Default: Goes to same directory as provided data</li>
<li><strong>alpha</strong>, The p-value threshold for calling sites as positively or negatively selected. Note that this argument has 0 bearing on JSON output. Default: 0.1</li>
<li><strong>genetic_code</strong>, the genetic code to use in codon analysis, Default: Universal. Consult NIH for details.</li>
<li><strong>bootstrap</strong>, The number of samples used to assess ancestral reconstruction uncertainty, in [0,100000]. Default:100.</li>
<li><strong>nexus</strong>, a Boolean <em>only required when</em> a nexus file is provided to the argument <cite>data</cite>. Default: False.</li>
</ol>
</dd>
</dl>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default SLAC analysis, where data is contained in a single file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myslac</span> <span class="o">=</span> <span class="n">SLAC</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default SLAC analysis, where data is contained in a single NEXUS file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myslac</span> <span class="o">=</span> <span class="n">SLAC</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.nex&quot;</span><span class="p">,</span> <span class="n">nexus</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default SLAC analysis, where alignment and tree are in separate files </span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myslac</span> <span class="o">=</span> <span class="n">SLAC</span><span class="p">(</span><span class="n">alignment</span> <span class="o">=</span> <span class="s2">&quot;/path/to/alignment.fasta&quot;</span><span class="p">,</span> <span class="n">tree</span> <span class="o">=</span> <span class="s2">&quot;/path/to/tree.tre&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a SLAC analysis, specifying that selection be tested only on leaves</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myslac</span> <span class="o">=</span> <span class="n">SLAC</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">branches</span> <span class="o">=</span> <span class="s2">&quot;Leaves&quot;</span> <span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a SLAC analysis, specifying 150 bootstrap replicates be used for ASR uncertainty</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myslac</span> <span class="o">=</span> <span class="n">SLAC</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">bootstrap</span> <span class="o">=</span> <span class="mi">150</span> <span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a SLAC analysis, specifying a custom JSON output file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myslac</span> <span class="o">=</span> <span class="n">SLAC</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;slac.json&quot;</span> <span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define SLAC analysis with a custom Hyphy, which is also defined here:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myhyphy</span> <span class="o">=</span> <span class="n">HyPhy</span><span class="p">(</span><span class="n">suppress_log</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">quiet</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="c1">## HyPhy will use default canonical install but run in full quiet mode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myslac</span> <span class="o">=</span> <span class="n">SLAC</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">hyphy</span><span class="o">=</span><span class="n">myhyphy</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Execute a defined SLAC instance</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myslac</span><span class="o">.</span><span class="n">run_analysis</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="analysis.ABSREL">
<em class="property">class </em><code class="descclassname">analysis.</code><code class="descname">ABSREL</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#analysis.ABSREL" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#analysis.Analysis" title="analysis.Analysis"><code class="xref py py-class docutils literal notranslate"><span class="pre">analysis.Analysis</span></code></a></p>
<dl class="docutils">
<dt>Required arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>alignment</strong> and <strong>tree</strong> OR <strong>data</strong>, either a file for alignment and tree separately, OR a file with both (combo FASTA/newick or nexus). Note that if a NEXUS file is provided with the <cite>data</cite> argument, the additional argument <cite>nexus=True</cite> must be supplied.</li>
</ol>
</dd>
<dt>Optional keyword arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>hyphy</strong>, a <code class="code docutils literal notranslate"><span class="pre">HyPhy()</span></code> instance. Default: Assumes canonical HyPhy install.</li>
<li><strong>branches</strong>, Branches to consider in site-level selection inference. Values “All”, “Internal”, “Leaves”, “Unlabeled branches”, or a <strong>specific label</strong> are accepted</li>
<li><strong>output</strong>, Name (and path to) to final output JSON file. Default: Goes to same directory as provided data</li>
<li><strong>genetic_code</strong>, the genetic code to use in codon analysis, Default: Universal. Consult NIH for details.</li>
<li><strong>nexus</strong>, a Boolean <em>only required when</em> a nexus file is provided to the argument <cite>data</cite>. Default: False.</li>
</ol>
</dd>
</dl>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default ABSREL analysis, where data is contained in a single file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myabsrel</span> <span class="o">=</span> <span class="n">ABSREL</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default ABSREL analysis, where data is contained in a single NEXUS file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myabsrel</span> <span class="o">=</span> <span class="n">ABSREL</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.nex&quot;</span><span class="p">,</span> <span class="n">nexus</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default ABSREL analysis, where alignment and tree are in separate files </span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myabsrel</span> <span class="o">=</span> <span class="n">ABSREL</span><span class="p">(</span><span class="n">alignment</span> <span class="o">=</span> <span class="s2">&quot;/path/to/alignment.fasta&quot;</span><span class="p">,</span> <span class="n">tree</span> <span class="o">=</span> <span class="s2">&quot;/path/to/tree.tre&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a ABSREL analysis, specifying that selection be tested only on leaves</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myabsrel</span> <span class="o">=</span> <span class="n">ABSREL</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">branches</span> <span class="o">=</span> <span class="s2">&quot;Leaves&quot;</span> <span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define ABSREL analysis with a custom Hyphy, which is also defined here:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myhyphy</span> <span class="o">=</span> <span class="n">HyPhy</span><span class="p">(</span><span class="n">suppress_log</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">quiet</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="c1">## HyPhy will use default canonical install but run in full quiet mode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myabsrel</span> <span class="o">=</span> <span class="n">ABSREL</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">hyphy</span><span class="o">=</span><span class="n">myhyphy</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Execute a defined ABSREL instance</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myabsrel</span><span class="o">.</span><span class="n">run_analysis</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="analysis.BUSTED">
<em class="property">class </em><code class="descclassname">analysis.</code><code class="descname">BUSTED</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#analysis.BUSTED" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#analysis.Analysis" title="analysis.Analysis"><code class="xref py py-class docutils literal notranslate"><span class="pre">analysis.Analysis</span></code></a></p>
<dl class="docutils">
<dt>Required arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>alignment</strong> and <strong>tree</strong> OR <strong>data</strong>, either a file for alignment and tree separately, OR a file with both (combo FASTA/newick or nexus). Note that if a NEXUS file is provided with the <cite>data</cite> argument, the additional argument <cite>nexus=True</cite> must be supplied.</li>
</ol>
</dd>
<dt>Optional keyword arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>hyphy</strong>, a <code class="code docutils literal notranslate"><span class="pre">HyPhy()</span></code> instance. Default: Assumes canonical HyPhy install.</li>
<li><strong>branches</strong>, Branches to consider in site-level selection inference. Values “All”, “Internal”, “Leaves”, “Unlabeled branches”, or a <strong>specific label</strong> are accepted</li>
<li><strong>output</strong>, Name (and path to) to final output JSON file. Default: Goes to same directory as provided data</li>
<li><strong>genetic_code</strong>, the genetic code to use in codon analysis, Default: Universal. Consult NIH for details.</li>
<li><strong>nexus</strong>, a Boolean <em>only required when</em> a nexus file is provided to the argument <cite>data</cite>. Default: False.</li>
</ol>
</dd>
</dl>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default BUSTED analysis, where data is contained in a single file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mybusted</span> <span class="o">=</span> <span class="n">BUSTED</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default BUSTED analysis, where data is contained in a single NEXUS file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mybusted</span> <span class="o">=</span> <span class="n">BUSTED</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.nex&quot;</span><span class="p">,</span> <span class="n">nexus</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default BUSTED analysis, where alignment and tree are in separate files </span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mybusted</span> <span class="o">=</span> <span class="n">BUSTED</span><span class="p">(</span><span class="n">alignment</span> <span class="o">=</span> <span class="s2">&quot;/path/to/alignment.fasta&quot;</span><span class="p">,</span> <span class="n">tree</span> <span class="o">=</span> <span class="s2">&quot;/path/to/tree.tre&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a BUSTED analysis, specifying that selection be tested only on leaves</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mybusted</span> <span class="o">=</span> <span class="n">BUSTED</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">branches</span> <span class="o">=</span> <span class="s2">&quot;Leaves&quot;</span> <span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define BUSTED analysis with a custom Hyphy, which is also defined here:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myhyphy</span> <span class="o">=</span> <span class="n">HyPhy</span><span class="p">(</span><span class="n">suppress_log</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">quiet</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="c1">## HyPhy will use default canonical install but run in full quiet mode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mybusted</span> <span class="o">=</span> <span class="n">BUSTED</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">hyphy</span><span class="o">=</span><span class="n">myhyphy</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Execute a defined BUSTED instance</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mybusted</span><span class="o">.</span><span class="n">run_analysis</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="analysis.RELAX">
<em class="property">class </em><code class="descclassname">analysis.</code><code class="descname">RELAX</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#analysis.RELAX" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#analysis.Analysis" title="analysis.Analysis"><code class="xref py py-class docutils literal notranslate"><span class="pre">analysis.Analysis</span></code></a></p>
<dl class="docutils">
<dt>Required arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>alignment</strong> and <strong>tree</strong> OR <strong>data</strong>, either a file for alignment and tree separately, OR a file with both (combo FASTA/newick or nexus). Note that if a NEXUS file is provided with the <cite>data</cite> argument, the additional argument <cite>nexus=True</cite> must be supplied.</li>
</ol>
</dd>
<dt>Optional keyword arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>hyphy</strong>, a <code class="code docutils literal notranslate"><span class="pre">HyPhy()</span></code> instance. Default: Assumes canonical HyPhy install.</li>
<li><strong>test_label</strong>, The label (must be found in your tree) corresponding to the <strong>test</strong> branch set</li>
<li><strong>reference_label</strong>, The label f(must be found in your tree) corresponding to the <strong>reference</strong> branch set. <strong>Only provide this argument if your tree has multiple labels in it.</strong></li>
<li><strong>output</strong>, Name (and path to) to final output JSON file. Default: Goes to same directory as provided data</li>
<li><strong>analysis_type</strong>, “All” (run hypothesis test and fit descriptive models) or “Minimal” (only run hypothesis test). Default: “All”.</li>
<li><strong>genetic_code</strong>, the genetic code to use in codon analysis, Default: Universal. Consult NIH for details.</li>
<li><strong>nexus</strong>, a Boolean <em>only required when</em> a nexus file is provided to the argument <cite>data</cite>. Default: False.</li>
</ol>
</dd>
</dl>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default RELAX analysis, where data is contained in a single file and test branches are labeled &quot;test&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myrelax</span> <span class="o">=</span> <span class="n">RELAX</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">test_label</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default RELAX analysis, where data is contained in a single NEXUS file and test branches are labeled &quot;test&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myrelax</span> <span class="o">=</span> <span class="n">RELAX</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.nex&quot;</span><span class="p">,</span> <span class="n">nexus</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">test_label</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default RELAX analysis, where alignment and tree are in separate files and test branches are labeled &quot;test&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myrelax</span> <span class="o">=</span> <span class="n">RELAX</span><span class="p">(</span><span class="n">alignment</span> <span class="o">=</span> <span class="s2">&quot;/path/to/alignment.fasta&quot;</span><span class="p">,</span> <span class="n">tree</span> <span class="o">=</span> <span class="s2">&quot;/path/to/tree.tre&quot;</span><span class="p">,</span> <span class="n">test_label</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default RELAX analysis, where data is contained in a single file, test branches are labeled &quot;test&quot;, and reference branches are labeled &quot;ref&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myrelax</span> <span class="o">=</span> <span class="n">RELAX</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">test_label</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span><span class="p">,</span> <span class="n">reference_label</span> <span class="o">=</span> <span class="s2">&quot;ref&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a default RELAX analysis, where data is contained in a single file, test branches are labeled &quot;test&quot;, and the Minimal analysis is run</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myrelax</span> <span class="o">=</span> <span class="n">RELAX</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">test_label</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span><span class="p">,</span> <span class="n">analysis_type</span> <span class="o">=</span> <span class="s2">&quot;Minimal&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define RELAX analysis with a custom Hyphy, which is also defined here:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myhyphy</span> <span class="o">=</span> <span class="n">HyPhy</span><span class="p">(</span><span class="n">suppress_log</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">quiet</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="c1">## HyPhy will use default canonical install but run in full quiet mode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myrelax</span> <span class="o">=</span> <span class="n">RELAX</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">hyphy</span><span class="o">=</span><span class="n">myhyphy</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Execute a defined RELAX instance</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myrelax</span><span class="o">.</span><span class="n">run_analysis</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="analysis.LEISR">
<em class="property">class </em><code class="descclassname">analysis.</code><code class="descname">LEISR</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#analysis.LEISR" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#analysis.Analysis" title="analysis.Analysis"><code class="xref py py-class docutils literal notranslate"><span class="pre">analysis.Analysis</span></code></a></p>
<dl class="docutils">
<dt>Required arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>alignment</strong> and <strong>tree</strong> OR <strong>data</strong>, either a file for alignment and tree separately, OR a file with both (combo FASTA/newick or nexus). Note that if a NEXUS file is provided with the <cite>data</cite> argument, the additional argument <cite>nexus=True</cite> must be supplied.</li>
<li><strong>type</strong>, either “nucleotide” or “protein” indicating the type of data being analyzed</li>
</ol>
</dd>
<dt>Optional keyword arguments:</dt>
<dd><ol class="first last arabic simple">
<li><strong>hyphy</strong>, a <code class="code docutils literal notranslate"><span class="pre">HyPhy()</span></code> instance. Default: Assumes canonical HyPhy install.</li>
<li><strong>model</strong>, The model to use to fit relative rates, i.e. GTR for nucleotides or LG for amino acids. For full options, please see HyPhy. Default: JC69.</li>
<li><strong>rate_variation</strong>, Whether to apply rate variation to branch length optimization. Options include No, Gamma, GDD. Note that Gamma and GDD will use four categories each. Default: No</li>
<li><strong>nexus</strong>, a Boolean <em>only required when</em> a nexus file is provided to the argument <cite>data</cite>. Default: False.</li>
</ol>
</dd>
</dl>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a LEISR Protein analysis, where data is contained in a single file and the WAG model with no rate variation is used</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myleisr</span> <span class="o">=</span> <span class="n">LEISR</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="nb">type</span> <span class="o">=</span> <span class="s2">&quot;protein&quot;</span><span class="p">,</span> <span class="n">model</span> <span class="o">=</span> <span class="s2">&quot;WAG&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a LEISR Protein analysis, where data is contained in a single NEXUS file and the WAG model with no rate variation is used</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myleisr</span> <span class="o">=</span> <span class="n">LEISR</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree..nex&quot;</span><span class="p">,</span> <span class="n">nexus</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">type</span> <span class="o">=</span> <span class="s2">&quot;protein&quot;</span><span class="p">,</span> <span class="n">model</span> <span class="o">=</span> <span class="s2">&quot;WAG&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a LEISR Protein analysis, where data is contained in a single file and the WAG model with Gamma rate variation is used</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myleisr</span> <span class="o">=</span> <span class="n">LEISR</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="nb">type</span> <span class="o">=</span> <span class="s2">&quot;protein&quot;</span><span class="p">,</span> <span class="n">model</span> <span class="o">=</span> <span class="s2">&quot;WAG&quot;</span><span class="p">,</span> <span class="n">rate_variation</span> <span class="o">=</span> <span class="s2">&quot;Gamma&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define a LEISR Nucleotide analysis, where data is contained in a single file and the HKY85 model with GDD rate variation is used</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myleisr</span> <span class="o">=</span> <span class="n">LEISR</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="nb">type</span> <span class="o">=</span> <span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="n">model</span> <span class="o">=</span> <span class="s2">&quot;HKY85&quot;</span><span class="p">,</span> <span class="n">rate_variation</span> <span class="o">=</span> <span class="s2">&quot;GDD&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Define LEISR analysis with a custom Hyphy, which is also defined here:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myhyphy</span> <span class="o">=</span> <span class="n">HyPhy</span><span class="p">(</span><span class="n">suppress_log</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">quiet</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="c1">## HyPhy will use default canonical install but run in full quiet mode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myleisr</span> <span class="o">=</span> <span class="n">LEISR</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;/path/to/data_with_tree.dat&quot;</span><span class="p">,</span> <span class="n">hyphy</span><span class="o">=</span><span class="n">myhyphy</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">### Execute a defined LEISR instance</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myleisr</span><span class="o">.</span><span class="n">run_analysis</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

</div>


      </div>
      <div class="bottomnav" role="navigation" aria-label="bottom navigation">
      
        <p>
        «&#160;&#160;<a href="hyphy.html"><code class="docutils literal notranslate"><span class="pre">hyphy</span></code> Module</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="extractor.html"><code class="docutils literal notranslate"><span class="pre">extractor</span></code> Module</a>&#160;&#160;»
        </p>

      </div>

    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Stephanie J. Spielman.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.2.
    </div>
  </body>
</html>